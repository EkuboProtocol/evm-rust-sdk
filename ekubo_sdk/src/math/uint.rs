use num_traits::Zero;

uint::construct_uint! {
    pub struct U256(4);
}

uint::construct_uint! {
    pub(crate) struct U512(8);
}

impl From<U256> for U512 {
    fn from(value: U256) -> Self {
        let mut result = U512::default();
        for (i, &limb) in value.0.iter().enumerate() {
            result.0[i] = limb;
        }
        result
    }
}

impl TryFrom<U512> for U256 {
    type Error = ();

    fn try_from(value: U512) -> Result<Self, Self::Error> {
        let mut result = U256::default();
        for (i, &limb) in value.0.iter().enumerate() {
            if i >= 4 && !limb.is_zero() {
                return Err(());
            }
            if i < 4 {
                result.0[i] = limb;
            }
        }
        Ok(result)
    }
}