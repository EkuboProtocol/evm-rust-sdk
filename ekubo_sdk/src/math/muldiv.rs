use crate::math::uint::{U256, U512};

#[derive(Debug, PartialEq)]
pub enum MuldivError {
    Overflow,
    DenominatorZero,
}

pub(crate) fn muldiv(x: U256, y: U256, d: U256, round_up: bool) -> Result<U256, MuldivError> {
    if d.is_zero() {
        return Err(MuldivError::DenominatorZero);
    }

    let intermediate: U512 = U512::from(x) * U512::from(y);
    let (quotient, remainder) = intermediate.div_mod(U512::from(d));

    let result = if round_up && !remainder.is_zero() {
        quotient + U512::one()
    } else {
        quotient
    };

    result.try_into().map_err(|_| MuldivError::Overflow)
}


#[cfg(test)]
mod tests {
    use super::*;
    use crate::math::uint::U256;

    #[test]
    fn test_muldiv_no_rounding() {
        // Test without rounding
        let x = U256::from(6);
        let y = U256::from(7);
        let d = U256::from(2);
        let result = muldiv(x, y, d, false).unwrap();
        assert_eq!(result, U256::from(21)); // (6 * 7) / 2 = 21
    }

    #[test]
    fn test_muldiv_with_rounding() {
        // Test with rounding up
        let x = U256::from(6);
        let y = U256::from(7);
        let d = U256::from(4);
        let result = muldiv(x, y, d, true).unwrap();
        assert_eq!(result, U256::from(11)); // (6 * 7) / 4 = 10.5, rounds up to 11
    }

    #[test]
    fn test_muldiv_no_rounding_needed() {
        // Test where rounding doesn't change the result
        let x = U256::from(8);
        let y = U256::from(2);
        let d = U256::from(4);
        let result = muldiv(x, y, d, true).unwrap();
        assert_eq!(result, U256::from(4)); // (8 * 2) / 4 = 4, exact division
    }

    #[test]
    fn test_muldiv_divide_by_zero() {
        // Test division by zero
        let x = U256::from(1);
        let y = U256::from(1);
        let d = U256::zero();
        let result = muldiv(x, y, d, false);
        assert!(matches!(result, Err(MuldivError::DenominatorZero)));
    }

    #[test]
    fn test_muldiv_overflow() {
        // Test overflow when result doesn't fit into U256
        let x = U256::max_value();
        let y = U256::from(2);
        let d = U256::from(1);
        let result = muldiv(x, y, d, false);
        assert!(matches!(result, Err(MuldivError::Overflow)));
    }

    #[test]
    fn test_muldiv_large_numbers() {
        // Test with large numbers that fit into U256
        let x = U256::from_dec_str("123456789012345678901234567890").unwrap();
        let y = U256::from_dec_str("987654321098765432109876543210").unwrap();
        let d = U256::from(1);
        let result = muldiv(x, y, d, false).unwrap();

        let expected = U256::from_dec_str("121932631137021795226185032733622923332237463801111263526900").unwrap();
        assert_eq!(result, expected);
    }

    #[test]
    fn test_muldiv_rounding_behavior() {
        // Test rounding behavior when remainder is zero
        let x = U256::from(10);
        let y = U256::from(10);
        let d = U256::from(5);
        let result = muldiv(x, y, d, true).unwrap();
        assert_eq!(result, U256::from(20)); // Exact division, no rounding needed

        // Test rounding behavior when remainder is not zero
        let d = U256::from(6);
        let result = muldiv(x, y, d, true).unwrap();
        assert_eq!(result, U256::from(17)); // (100 / 6) = 16.666..., rounds up to 17
    }

    #[test]
    fn test_muldiv_zero_numerator() {
        // Test with zero numerator
        let x = U256::zero();
        let y = U256::from(100);
        let d = U256::from(10);
        let result = muldiv(x, y, d, false).unwrap();
        assert_eq!(result, U256::zero());
    }

    #[test]
    fn test_muldiv_one_denominator() {
        // Test with denominator of one
        let x = U256::from(123456789);
        let y = U256::from(987654321);
        let d = U256::one();
        let result = muldiv(x, y, d, false).unwrap();
        let expected = U256::from(121932631112635269u64);
        assert_eq!(result, expected);
    }
}
