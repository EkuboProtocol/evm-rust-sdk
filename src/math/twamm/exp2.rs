use crate::math::uint::U256;

pub fn exp2(x: u128) -> u128 {
    // x must be less than 0x400000000000000000 == 64 << 64 (overflow check)
    assert!(x < 0x400000000000000000, "Overflow");

    // Start with 0x80000000000000000000000000000000 (i.e. 2^127)
    let mut result = U256::one() << 127;

    macro_rules! mul_shift {
        ($mask:expr, $factor:expr) => {
            if (x & $mask) != 0 {
                result = (result * $factor) >> 128;
            }
        };
    }

    // Each branch multiplies result by a precomputed constant if the corresponding bit is set.
    mul_shift!(
        0x8000000000000000,
        U256([12896923290648804670, 7640891576956012808, 1, 0])
    );
    mul_shift!(
        0x4000000000000000,
        U256([10185530680776453612, 3490255227380126430, 1, 0])
    );
    mul_shift!(
        0x2000000000000000,
        U256([17854737389604993311, 1669572981167730125, 1, 0])
    );
    mul_shift!(
        0x1000000000000000,
        U256([10057302232697242467, 816707133613602345, 1, 0])
    );
    mul_shift!(
        0x800000000000000,
        U256([8958942473441515517, 403931097166463918, 1, 0])
    );
    mul_shift!(
        0x400000000000000,
        U256([17855306115364535784, 200871872941133542, 1, 0])
    );
    mul_shift!(
        0x200000000000000,
        U256([5361163942339934783, 100163996173424344, 1, 0])
    );
    mul_shift!(
        0x100000000000000,
        U256([3002515481205445955, 50014196964519265, 1, 0])
    );
    mul_shift!(
        0x80000000000000,
        U256([18381954890662373275, 24990171141283489, 1, 0])
    );
    mul_shift!(
        0x40000000000000,
        U256([7863777292665987087, 12490856599448656, 1, 0])
    );
    mul_shift!(
        0x20000000000000,
        U256([2935407722259146207, 6244371414720417, 1, 0])
    );
    mul_shift!(
        0x10000000000000,
        U256([4099144763823068965, 3121921530820282, 1, 0])
    );
    mul_shift!(
        0x8000000000000,
        U256([7830867710921767997, 1560894726863213, 1, 0])
    );
    mul_shift!(
        0x4000000000000,
        U256([15030865137112770483, 780430854493329, 1, 0])
    );
    mul_shift!(
        0x2000000000000,
        U256([6072397496287043622, 390211300099399, 1, 0])
    );
    mul_shift!(
        0x1000000000000,
        U256([439311289473205244, 195104618273796, 1, 0])
    );
    mul_shift!(
        0x800000000000,
        U256([3381501736036650923, 97552051194286, 1, 0])
    );
    mul_shift!(
        0x400000000000,
        U256([11878183186495682505, 48775961111660, 1, 0])
    );
    mul_shift!(
        0x200000000000,
        U256([436289341605138797, 24387964434481, 1, 0])
    );
    mul_shift!(
        0x100000000000,
        U256([15706463035261523811, 12193978186905, 1, 0])
    );
    mul_shift!(
        0x80000000000,
        U256([10951006159863813672, 6096988085869, 1, 0])
    );
    mul_shift!(
        0x40000000000,
        U256([100320500613761579, 3048493791039, 1, 0])
    );
    mul_shift!(
        0x20000000000,
        U256([10331567486136386213, 1524246832545, 1, 0])
    );
    mul_shift!(
        0x10000000000,
        U256([5442291021268196350, 762123400529, 1, 0])
    );
    mul_shift!(
        0x8000000000,
        U256([14320987186005390063, 381061696328, 1, 0])
    );
    mul_shift!(
        0x4000000000,
        U256([7754798478847538042, 190530847180, 1, 0])
    );
    mul_shift!(0x2000000000, U256([4025998895411732230, 95265423344, 1, 0]));
    mul_shift!(
        0x1000000000,
        U256([11273524327873463545, 47632711610, 1, 0])
    );
    mul_shift!(0x800000000, U256([17175265786997884963, 23816355789, 1, 0]));
    mul_shift!(0x400000000, U256([2248886808180014672, 11908177891, 1, 0]));
    mul_shift!(0x200000000, U256([11068971934550065103, 5954088944, 1, 0]));
    mul_shift!(0x100000000, U256([1103089072964130132, 2977044472, 1, 0]));
    mul_shift!(0x80000000, U256([17890439386703286701, 1488522235, 1, 0]));
    mul_shift!(0x40000000, U256([17891629424345511116, 744261117, 1, 0]));
    mul_shift!(0x20000000, U256([18099946172563701162, 372130558, 1, 0]));
    mul_shift!(0x10000000, U256([9032662942166067630, 186065279, 1, 0]));
    mul_shift!(0x8000000, U256([13735375971908885710, 93032639, 1, 0]));
    mul_shift!(0x4000000, U256([16089978138801990413, 46516319, 1, 0]));
    mul_shift!(0x2000000, U256([17268090635253964293, 23258159, 1, 0]));
    mul_shift!(0x1000000, U256([17857349736731306316, 11629079, 1, 0]));
    mul_shift!(0x800000, U256([18152030000782816061, 5814539, 1, 0]));
    mul_shift!(0x400000, U256([18299382811136780613, 2907269, 1, 0]));
    mul_shift!(0x200000, U256([18373062385895815308, 1453634, 1, 0]));
    mul_shift!(0x100000, U256([9186530928816069952, 726817, 1, 0]));
    mul_shift!(0x80000, U256([13816637435229851358, 363408, 1, 0]));
    mul_shift!(0x40000, U256([6908318701106685822, 181704, 1, 0]));
    mul_shift!(0x20000, U256([3454159346426282946, 90852, 1, 0]));
    mul_shift!(0x10000, U256([1727079672181376481, 45426, 1, 0]));
    mul_shift!(0x8000, U256([863539835832746992, 22713, 1, 0]));
    mul_shift!(0x4000, U256([9655141954706663992, 11356, 1, 0]));
    mul_shift!(0x2000, U256([4827570977337210668, 5678, 1, 0]));
    mul_shift!(0x1000, U256([2413785488664575002, 2839, 1, 0]));
    mul_shift!(0x800, U256([10430264781186055726, 1419, 1, 0]));
    mul_shift!(0x400, U256([14438504427447551775, 709, 1, 0]));
    mul_shift!(0x200, U256([16442624250578488721, 354, 1, 0]));
    mul_shift!(0x100, U256([8221312125289228617, 177, 1, 0]));
    mul_shift!(0x80, U256([13334028099499386180, 88, 1, 0]));
    mul_shift!(0x40, U256([6667014049749692106, 44, 1, 0]));
    mul_shift!(0x20, U256([3333507024874845807, 22, 1, 0]));
    mul_shift!(0x10, U256([1666753512437422842, 11, 1, 0]));
    mul_shift!(0x8, U256([10056748793073487213, 5, 1, 0]));
    mul_shift!(0x4, U256([14251746433391519410, 2, 1, 0]));
    mul_shift!(0x2, U256([7125873216695759704, 1, 1, 0]));
    mul_shift!(0x1, U256([12786308645202655659, 0, 1, 0]));

    // Final adjustment: shift right by 63 - (x >> 64). (x >> 64) is the integer part.
    result >>= 63 - (x >> 64);

    result.as_u128()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_exp2_cases() {
        assert_eq!(exp2(0), 1 << 64);
        assert_eq!(exp2(1 << 64), 2 << 64);
        assert_eq!(exp2(2 << 64), 4 << 64);
        assert_eq!(exp2(3 << 64), 8 << 64);
        assert_eq!(exp2(63 << 64), 9223372036854775808 << 64);
    }
}
