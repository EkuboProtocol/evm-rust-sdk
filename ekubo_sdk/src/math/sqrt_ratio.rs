use crate::math::muldiv::{muldiv, MuldivError};
use crate::math::uint::U256;
use num_traits::Zero;

#[derive(Debug, PartialEq)]
pub enum PriceMathError {
    NoLiquidity,
    Overflow,
    InvalidDenominator,
    NegativeResult,
    ExceedsMaxValue,
    MuldivError(MuldivError),
}

pub fn next_sqrt_ratio_from_amount0(
    sqrt_ratio: U256,
    liquidity: u128,
    amount0: i128,
) -> Result<U256, PriceMathError> {
    if amount0 == 0 {
        return Ok(sqrt_ratio);
    }

    if liquidity.is_zero() {
        return Err(PriceMathError::NoLiquidity);
    }

    let numerator1 = U256::from(liquidity) << 128;

    if amount0 < 0 {
        // amount0 is negative
        let amount0_abs = U256::from(amount0.checked_abs().ok_or(PriceMathError::Overflow)?);

        // product = amount0_abs * sqrt_ratio
        let product = amount0_abs.checked_mul(sqrt_ratio).ok_or(PriceMathError::Overflow)?;

        let denominator = numerator1.checked_sub(product).ok_or(PriceMathError::Overflow)?;

        muldiv(numerator1, sqrt_ratio, denominator, true).map_err(PriceMathError::MuldivError)
    } else {
        // amount0 is positive
        let amount0_u256 = U256::from(amount0);

        let denom_p1 = numerator1 / sqrt_ratio;

        let denom = denom_p1
            .checked_add(amount0_u256)
            .ok_or(PriceMathError::Overflow)?;

        muldiv(numerator1, U256::one(), denom, true)
            .map_err(PriceMathError::MuldivError)
    }
}

const TWO_POW_128: U256 = U256([0, 0, 1, 0]);

pub fn next_sqrt_ratio_from_amount1(
    sqrt_ratio: U256,
    liquidity: u128,
    amount1: i128,
) -> Result<U256, PriceMathError> {
    if amount1 == 0 {
        return Ok(sqrt_ratio);
    }

    if liquidity.is_zero() {
        return Err(PriceMathError::NoLiquidity);
    }

    let amount1_abs = U256::from(amount1.abs());

    let round_up = amount1 < 0;

    let quotient = muldiv(amount1_abs, TWO_POW_128, liquidity.into(), round_up)
        .map_err(PriceMathError::MuldivError)?;

    if amount1 < 0 {
        sqrt_ratio
            .checked_sub(quotient)
            .ok_or(PriceMathError::NegativeResult)
    } else {
        sqrt_ratio
            .checked_add(quotient)
            .ok_or(PriceMathError::Overflow)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_next_sqrt_ratio_from_amount0_add_price_goes_down() {
        // Corresponds to:
        // nextSqrtRatioFromAmount0(1n << 128n, 1000000n, 1000n)
        let sqrt_ratio = U256::from(1u64) << 128;
        let liquidity = 1_000_000u128;
        let amount0 = 1000i128;
        let result = next_sqrt_ratio_from_amount0(sqrt_ratio, liquidity, amount0).unwrap();
        let expected = U256::from_dec_str("339942424496442021441932674757011200256").unwrap();
        assert_eq!(result, expected);
    }

    #[test]
    fn test_next_sqrt_ratio_from_amount0_exact_out_overflow() {
        // Corresponds to:
        // nextSqrtRatioFromAmount0(1n << 128n, 1n, -100000000000000n)
        let sqrt_ratio = U256::from(1u64) << 128;
        let liquidity = 1u128;
        let amount0 = -100_000_000_000_000i128;
        let result = next_sqrt_ratio_from_amount0(sqrt_ratio, liquidity, amount0);
        assert!(result.is_err());
        assert_eq!(result.err().unwrap(), PriceMathError::Overflow);
    }

    #[test]
    fn test_next_sqrt_ratio_from_amount0_exact_in_cant_underflow() {
        // Corresponds to:
        // nextSqrtRatioFromAmount0(1n << 128n, 1n, 100000000000000n)
        let sqrt_ratio = U256::from(1u64) << 128;
        let liquidity = 1u128;
        let amount0 = 100_000_000_000_000i128;
        let result = next_sqrt_ratio_from_amount0(sqrt_ratio, liquidity, amount0).unwrap();
        let expected = U256::from_dec_str("3402823669209350606397054").unwrap();
        assert_eq!(result, expected);
    }

    #[test]
    fn test_next_sqrt_ratio_from_amount0_sub_price_goes_up() {
        // Corresponds to:
        // nextSqrtRatioFromAmount0(1n << 128n, 100000000000n, -1000n)
        let sqrt_ratio = U256::from(1u64) << 128;
        let liquidity = 100_000_000_000u128;
        let amount0 = -1000i128;
        let result = next_sqrt_ratio_from_amount0(sqrt_ratio, liquidity, amount0).unwrap();
        let expected =
            U256::from_dec_str("340282370323762166700996274441730955874").unwrap();
        assert_eq!(result, expected);
    }

    #[test]
    fn test_next_sqrt_ratio_from_amount1_add_price_goes_up() {
        // Corresponds to:
        // nextSqrtRatioFromAmount1(1n << 128n, 1000000n, 1000n)
        let sqrt_ratio = U256::from(1u64) << 128;
        let liquidity = 1_000_000u128;
        let amount1 = 1000i128;
        let result = next_sqrt_ratio_from_amount1(sqrt_ratio, liquidity, amount1).unwrap();
        let expected = U256::from_dec_str("340622649287859401926837982039199979667").unwrap();
        assert_eq!(result, expected);
    }

    #[test]
    fn test_next_sqrt_ratio_from_amount1_exact_out_overflow() {
        // Corresponds to:
        // nextSqrtRatioFromAmount1(1n << 128n, 1n, -100000000000000n)
        let sqrt_ratio = U256::from(1u64) << 128;
        let liquidity = 1u128;
        let amount1 = -100_000_000_000_000i128;
        let result = next_sqrt_ratio_from_amount1(sqrt_ratio, liquidity, amount1);
        assert!(result.is_err());
        assert_eq!(result.err().unwrap(), PriceMathError::NegativeResult);
    }

    #[test]
    fn test_next_sqrt_ratio_from_amount1_exact_in_cant_underflow() {
        // Corresponds to:
        // nextSqrtRatioFromAmount1(1n << 128n, 1n, 100000000000000n)
        let sqrt_ratio = U256::from(1u64) << 128;
        let liquidity = 1u128;
        let amount1 = 100_000_000_000_000i128;
        let result = next_sqrt_ratio_from_amount1(sqrt_ratio, liquidity, amount1).unwrap();
        let expected = U256::from_dec_str("34028236692094186628704381681640284520207431768211456").unwrap();
        assert_eq!(result, expected);
    }

    #[test]
    fn test_next_sqrt_ratio_from_amount1_sub_price_goes_down() {
        // Corresponds to:
        // nextSqrtRatioFromAmount1(1n << 128n, 100000000000n, -1000n)
        let sqrt_ratio = U256::from(1u64) << 128;
        let liquidity = 100_000_000_000u128;
        let amount1 = -1000i128;
        let result = next_sqrt_ratio_from_amount1(sqrt_ratio, liquidity, amount1).unwrap();
        let expected =
            U256::from_dec_str("340282363518114794253989972798022137138").unwrap();
        assert_eq!(result, expected);
    }
}

