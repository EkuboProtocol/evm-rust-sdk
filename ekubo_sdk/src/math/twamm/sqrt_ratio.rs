use crate::math::muldiv::muldiv;
use crate::math::twamm::exp::exp;
use crate::math::uint::U256;
use num_traits::Zero;

const EXPONENT_LIMIT: U256 = U256([0, 88, 0, 0]);
const TWO_POW_128: U256 = U256([0, 0, 1, 0]);

pub fn calculate_next_sqrt_ratio(
    sqrt_ratio: U256,
    liquidity: u128,
    token0_sale_rate: u128,
    token1_sale_rate: u128,
    time_elapsed: u32,
    fee: u128,
) -> Option<U256> {
    let sqrt_sale_ratio =
        ((U256::from(token1_sale_rate) << 128) / token0_sale_rate).integer_sqrt() << 64;

    if liquidity.is_zero() {
        return Some(sqrt_sale_ratio);
    }

    let sale_rate = ((U256::from(token1_sale_rate) * U256::from(token0_sale_rate)).integer_sqrt()
        * (TWO_POW_128 - fee))
        / TWO_POW_128;

    let round_up = sqrt_ratio > sqrt_sale_ratio;

    let exponent = muldiv(
        sale_rate * U256::from(0x200000000u128),
        U256::from(time_elapsed),
        U256::from(liquidity),
        round_up,
    )
        .ok()?;

    if exponent > EXPONENT_LIMIT {
        return Some(sqrt_sale_ratio);
    }

    let e = exp(exponent.low_u128())?;

    let (num, sign) = if round_up {
        (sqrt_ratio - sqrt_sale_ratio, true)
    } else {
        (sqrt_sale_ratio - sqrt_ratio, false)
    };

    let c = muldiv(num, TWO_POW_128, sqrt_sale_ratio + sqrt_ratio, round_up).ok()?;

    let (term1, term2) = (e - c, e + c);

    let scale = if sign {
        muldiv(term2, TWO_POW_128, term1, round_up)
    } else {
        muldiv(term1, TWO_POW_128, term2, round_up)
    }
        .ok()?;

    muldiv(sqrt_sale_ratio, scale, TWO_POW_128, false).ok()
}

#[cfg(test)]
mod tests {
    use crate::math::twamm::sqrt_ratio::calculate_next_sqrt_ratio;
    use crate::math::uint::U256;
    use alloc::vec;

    const ONE_E18: u128 = 1_000_000_000_000_000_000; // 10^18
    const SHIFT_32: u128 = 1u128 << 32; // 2^32 = 4294967296
    const TOKEN_SALE_RATE: u128 = ONE_E18 * SHIFT_32; // 10^18 * 2^32

    struct TestCase {
        description: &'static str,
        sqrt_ratio: U256,
        liquidity: u128,
        token0_sale_rate: u128,
        token1_sale_rate: u128,
        time_elapsed: u32,
        fee: u128,
        expected: U256,
    }

    #[test]
    fn test_calculate_next_sqrt_ratio() {
        let test_cases = vec![
            TestCase {
                description: "liquidity is zero, price is sqrtSaleRatio",
                sqrt_ratio: U256::zero(),
                liquidity: 0,
                token0_sale_rate: TOKEN_SALE_RATE,
                token1_sale_rate: TOKEN_SALE_RATE,
                time_elapsed: 0,
                fee: 0,
                expected: U256::from_dec_str("340282366920938463463374607431768211456").unwrap(),
            },
            TestCase {
                description: "large exponent (> 88), price is sqrtSaleRatio",
                sqrt_ratio: U256::one() << 128,
                liquidity: 1,
                token0_sale_rate: TOKEN_SALE_RATE,
                token1_sale_rate: 1980 * ONE_E18 * SHIFT_32,
                time_elapsed: 1,
                fee: 0,
                expected: U256::from_dec_str("15141609448466370575819539296664158208000").unwrap(),
            },
            TestCase {
                description: "low liquidity, same sale rate",
                sqrt_ratio: U256::from(2u128) << 128,
                liquidity: 1,
                token0_sale_rate: TOKEN_SALE_RATE,
                token1_sale_rate: TOKEN_SALE_RATE,
                time_elapsed: 1,
                fee: 0,
                expected: U256::from_dec_str("340282366920938463463374607431768211456").unwrap(),
            },
            TestCase {
                description: "low liquidity, token0SaleRate > token1SaleRate",
                sqrt_ratio: U256::one() << 128,
                liquidity: 1,
                token0_sale_rate: 2 * TOKEN_SALE_RATE,
                token1_sale_rate: TOKEN_SALE_RATE,
                time_elapsed: 16,
                fee: 0,
                expected: U256::from_dec_str("240615969168004511538585310832300654592").unwrap(),
            },
            TestCase {
                description: "low liquidity, token1SaleRate > token0SaleRate",
                sqrt_ratio: U256::one() << 128,
                liquidity: 1,
                token0_sale_rate: TOKEN_SALE_RATE,
                token1_sale_rate: 2 * TOKEN_SALE_RATE,
                time_elapsed: 16,
                fee: 0,
                expected: U256::from_dec_str("481231938336009023077170621664601309184").unwrap(),
            },
            TestCase {
                description: "high liquidity, same sale rate",
                sqrt_ratio: U256::from(2u128) << 128,
                liquidity: 1_000_000 * ONE_E18,
                token0_sale_rate: TOKEN_SALE_RATE,
                token1_sale_rate: TOKEN_SALE_RATE,
                time_elapsed: 1,
                fee: 0,
                expected: U256::from_dec_str("680563712996817854544971649595310676716").unwrap(),
            },
            TestCase {
                description: "high liquidity, token0SaleRate > token1SaleRate",
                sqrt_ratio: U256::one() << 128,
                liquidity: 1_000_000 * ONE_E18,
                token0_sale_rate: 2 * TOKEN_SALE_RATE,
                token1_sale_rate: TOKEN_SALE_RATE,
                time_elapsed: 1,
                fee: 0,
                expected: U256::from_dec_str("340282026639252106118798709911530476351").unwrap(),
            },
            TestCase {
                description: "high liquidity, token1SaleRate > token0SaleRate",
                sqrt_ratio: U256::one() << 128,
                liquidity: 1_000_000 * ONE_E18,
                token0_sale_rate: TOKEN_SALE_RATE,
                token1_sale_rate: 2 * TOKEN_SALE_RATE,
                time_elapsed: 1,
                fee: 0,
                expected: U256::from_dec_str("340282707202965102147504331693640508329").unwrap(),
            },
            TestCase {
                description: "round in direction of price",
                sqrt_ratio: U256::from_dec_str("481231811499356508086519009265716982182").unwrap(),
                liquidity: 70_710_696_755_630_728_101_718_334,
                token0_sale_rate: 10_526_880_627_450_980_392_156_862_745,
                token1_sale_rate: 10_526_880_627_450_980_392_156_862_745,
                time_elapsed: 2040,
                fee: 0,
                expected: U256::from_dec_str("481207752340103616358571818546900413164").unwrap(),
            },
        ];

        for test_case in test_cases {
            let result = calculate_next_sqrt_ratio(
                test_case.sqrt_ratio,
                test_case.liquidity,
                test_case.token0_sale_rate,
                test_case.token1_sale_rate,
                test_case.time_elapsed,
                test_case.fee,
            )
                .unwrap(); // Assuming the function returns Option<U256>

            assert_eq!(
                result, test_case.expected,
                "Test case failed: {}",
                test_case.description
            );
        }
    }
}
