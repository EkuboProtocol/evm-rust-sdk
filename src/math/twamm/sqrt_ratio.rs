use crate::math::muldiv::muldiv;
use crate::math::twamm::exp2::exp2;
use crate::math::uint::U256;
use num_traits::Zero;

const EXPONENT_LIMIT: U256 = U256([0, 64, 0, 0]);
const TWO_POW_64: U256 = U256([0, 1, 0, 0]);

pub fn calculate_next_sqrt_ratio(
    sqrt_ratio: U256,
    liquidity: u128,
    token0_sale_rate: u128,
    token1_sale_rate: u128,
    time_elapsed: u32,
    fee: u64,
) -> Option<U256> {
    let sqrt_sale_ratio =
        ((U256::from(token1_sale_rate) << 128) / token0_sale_rate).integer_sqrt() << 64;

    if liquidity.is_zero() {
        return Some(sqrt_sale_ratio);
    }

    let sale_rate = ((U256::from(token1_sale_rate) * U256::from(token0_sale_rate)).integer_sqrt()
        * (TWO_POW_64 - fee))
        / TWO_POW_64;

    let round_up = sqrt_ratio > sqrt_sale_ratio;

    let exponent = muldiv(
        // 12392656037 = Math.floor(Math.LOG2E * 2**33).
        // this combines the doubling, the left shifting and the converting to a base 2 exponent into a single multiplication
        sale_rate * U256::from(12392656037i64),
        U256::from(time_elapsed),
        U256::from(liquidity),
        round_up,
    )
    .ok()?
    .low_u128();

    if exponent >= 0x400000000000000000 {
        return Some(sqrt_sale_ratio);
    }

    let e = U256::from(exp2(exponent.try_into().ok()?)) << 64;

    let (num, sign) = if round_up {
        (sqrt_ratio - sqrt_sale_ratio, true)
    } else {
        (sqrt_sale_ratio - sqrt_ratio, false)
    };

    let c = muldiv(
        num,
        TWO_POW_64 * TWO_POW_64,
        sqrt_sale_ratio + sqrt_ratio,
        round_up,
    )
    .ok()?;

    let (term1, term2) = (e - c, e + c);

    let scale = if sign {
        muldiv(term2, TWO_POW_64 * TWO_POW_64, term1, round_up)
    } else {
        muldiv(term1, TWO_POW_64 * TWO_POW_64, term2, round_up)
    }
    .ok()?;

    muldiv(sqrt_sale_ratio, scale, TWO_POW_64 * TWO_POW_64, false).ok()
}

#[cfg(test)]
mod tests {
    use crate::math::twamm::sqrt_ratio::calculate_next_sqrt_ratio;
    use crate::math::uint::U256;
    use alloc::vec;
    use insta::{assert_debug_snapshot, assert_snapshot};

    const ONE_E18: u128 = 1_000_000_000_000_000_000; // 10^18
    const SHIFT_32: u128 = 1u128 << 32; // 2^32 = 4294967296
    const TOKEN_SALE_RATE: u128 = ONE_E18 * SHIFT_32; // 10^18 * 2^32

    struct TestCase {
        description: &'static str,
        sqrt_ratio: U256,
        liquidity: u128,
        token0_sale_rate: u128,
        token1_sale_rate: u128,
        time_elapsed: u32,
        fee: u64,
    }

    #[test]
    fn test_calculate_next_sqrt_ratio() {
        let test_cases = vec![
            TestCase {
                description: "zero_liquidity_price_eq_sale_ratio",
                sqrt_ratio: U256::zero(),
                liquidity: 0,
                token0_sale_rate: TOKEN_SALE_RATE,
                token1_sale_rate: TOKEN_SALE_RATE,
                time_elapsed: 0,
                fee: 0,
            },
            TestCase {
                description: "large_exponent_price_sqrt_ratio",
                sqrt_ratio: U256::one() << 128,
                liquidity: 1,
                token0_sale_rate: TOKEN_SALE_RATE,
                token1_sale_rate: 1980 * ONE_E18 * SHIFT_32,
                time_elapsed: 1,
                fee: 0,
            },
            TestCase {
                description: "low_liquiidty_same_sale_ratio",
                sqrt_ratio: U256::from(2u128) << 128,
                liquidity: 1,
                token0_sale_rate: TOKEN_SALE_RATE,
                token1_sale_rate: TOKEN_SALE_RATE,
                time_elapsed: 1,
                fee: 0,
            },
            TestCase {
                description: "low_liquidity_token0_gt_token1",
                sqrt_ratio: U256::one() << 128,
                liquidity: 1,
                token0_sale_rate: 2 * TOKEN_SALE_RATE,
                token1_sale_rate: TOKEN_SALE_RATE,
                time_elapsed: 16,
                fee: 0,
            },
            TestCase {
                description: "low_liquidity_token1_gt_token0",
                sqrt_ratio: U256::one() << 128,
                liquidity: 1,
                token0_sale_rate: TOKEN_SALE_RATE,
                token1_sale_rate: 2 * TOKEN_SALE_RATE,
                time_elapsed: 16,
                fee: 0,
            },
            TestCase {
                description: "high_liquidity_same_sale_rate",
                sqrt_ratio: U256::from(2u128) << 128,
                liquidity: 1_000_000 * ONE_E18,
                token0_sale_rate: TOKEN_SALE_RATE,
                token1_sale_rate: TOKEN_SALE_RATE,
                time_elapsed: 1,
                fee: 0,
            },
            TestCase {
                description: "high_liquidity_token0_gt_token1",
                sqrt_ratio: U256::one() << 128,
                liquidity: 1_000_000 * ONE_E18,
                token0_sale_rate: 2 * TOKEN_SALE_RATE,
                token1_sale_rate: TOKEN_SALE_RATE,
                time_elapsed: 1,
                fee: 0,
            },
            TestCase {
                description: "high_liquidity_token1_gt_token0",
                sqrt_ratio: U256::one() << 128,
                liquidity: 1_000_000 * ONE_E18,
                token0_sale_rate: TOKEN_SALE_RATE,
                token1_sale_rate: 2 * TOKEN_SALE_RATE,
                time_elapsed: 1,
                fee: 0,
            },
            TestCase {
                description: "round_in_direction_of_price",
                sqrt_ratio: U256::from_dec_str("481231811499356508086519009265716982182").unwrap(),
                liquidity: 70_710_696_755_630_728_101_718_334,
                token0_sale_rate: 10_526_880_627_450_980_392_156_862_745,
                token1_sale_rate: 10_526_880_627_450_980_392_156_862_745,
                time_elapsed: 2040,
                fee: 0,
            },
        ];

        for test_case in test_cases {
            assert_debug_snapshot!(
                test_case.description,
                calculate_next_sqrt_ratio(
                    test_case.sqrt_ratio,
                    test_case.liquidity,
                    test_case.token0_sale_rate,
                    test_case.token1_sale_rate,
                    test_case.time_elapsed,
                    test_case.fee,
                )
            );
        }
    }
}
