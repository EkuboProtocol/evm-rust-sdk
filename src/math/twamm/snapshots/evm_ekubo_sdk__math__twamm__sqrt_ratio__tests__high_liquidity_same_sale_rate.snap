---
source: src/math/twamm/sqrt_ratio.rs
expression: "calculate_next_sqrt_ratio(test_case.sqrt_ratio, test_case.liquidity,\ntest_case.token0_sale_rate, test_case.token1_sale_rate,\ntest_case.time_elapsed, test_case.fee,)"
---
680563712996817890757827685335626524191
